macros/generic_tests/recency.sql

```diff
@@ -4,7 +4,21 @@
 
 {% macro default__test_recency(model, field, datepart, interval, group_by_columns) %}
 
-{% set threshold = dbt.dateadd(datepart, interval * -1, current_timestamp_backcompat()) %}
+{#- Get columns in model #}
+{%- set columns = adapter.get_columns_in_relation(model) %}
+
+{#- Get column type of field in model #}
+{%- set column_type = columns | selectattr("name", "==", field) | map(attribute="data_type") | first %}
+
+{#- Use a date-based threshold if column type is DATE or datepart is day or greater #}
+{%- if column_type == 'DATE' and datepart in ['year', 'quarter', 'month', 'week', 'day'] %}
+  {%- set threshold = dbt.dateadd(datepart, interval * -1, dbt.safe_cast(dbt.current_timestamp_backcompat(), api.Column.translate_type("date"))) %}
+{%- else %}
+  {%- set threshold = dbt.dateadd(datepart, interval * -1, dbt.current_timestamp_backcompat()) %}
+{%- endif %}
+
+
+{%- set field_datetime = dbt.safe_cast(field, api.Column.translate_type("datetime")) %}
```

Comment:
This doesn't seem to be used anywhere?

Reply:
You're right! Missed this from a previous iteration of a solution, thanks for catching it.